Add a configuration to aggregate automatically projects from reactor (no need to declare manually all depdencies)

olamy opened the pull request:
- **Add option to aggregate automatically reports from aggregators, no need to add any dependency**
- **make server.id configurable for snapshots deploy**
- **not needed**
- **spotless**

olamy commented:
@Godin Hi any chance to get this merged and released?

olamy commented on .azure-pipelines/azure-pipelines.yml:
```diff
@@ -80,7 +80,7 @@ jobs:
       set -e
       mkdir .maven
-      curl -L "https://archive.apache.org/dist/maven/maven-3/3.8.3/binaries/apache-maven-3.8.3-bin.tar.gz" -o .maven/maven.tar.gz
+      curl -L "https://repo.maven.apache.org/maven2/org/apache/maven/apache-maven/3.8.3/apache-maven-3.8.3-bin.tar.gz" -o .maven/maven.tar.gz
```
not really related. but this is a very practice to use Apache archive for this.

olamy commented on org.jacoco.examples.test/pom.xml:
```diff
@@ -38,6 +38,11 @@
       <groupId>${project.groupId}</groupId>
       <artifactId>org.jacoco.agent.rt</artifactId>
     </dependency>
+    <dependency>
```
with recent version of Maven this can be build before the maven plugin, so need to be sure the maven plugin is build first.

sbordet commented on jacoco-maven-plugin/src/org/jacoco/maven/ReportAggregateMojo.java:
```diff
@@ -91,6 +92,14 @@ public class ReportAggregateMojo extends AbstractReportMojo {
+	 * @since 0.8.13
+	 */
+	@Parameter(defaultValue = "false", property = "jacoco.onlyReactorProjects")
+	private boolean onlyReactorProjects;
```
Why is this property called `only...` and not `all...`?

From the javadoc seems more appropriate to call it `allReactorProjects`, or perhaps `allModules` or `allSubModules`?

olamy replied:
I was not really sure because the code currently reads only all dependencies declared in the coverage/aggregator project (dependencies that are part of the reactor). 
If the flag is activated, the code will not use any current protect dependencies but only(All :))ReactorProjects. 
So yes, something like `allReactorProjects` would also work, except it doesn't tell the user the configuration no longer reads the project's dependencies.
Doh, this looks complicated to explain so that it might be wrong :).

no real strong opinion. @Godin what do you think?

olamy commented:
@Godin ping

olamy commented:
@Godin any chance to have this merged or at least some comments on why this cannot be merged? 
Thanks

cpfeiffer commented:
Thanks a lot, we had a similar thing in mind. However, I would like to suggest creating a new mojo for this, and making it an aggregator mojo (`aggregator = true`).

Reason: when adding the goal to a maven command like this:
```
mvn clean
mvn verify jacoco:report-aggregate
```
you will successfully get a toplevel jacoco-report that includes all modules, however with no content at all. This is because
non-aggregator mojos are integrated into the normal build lifecycle. So the report for the toplevel module will be created before any coverage has been calculated for the modules.

Even worse, if you call 
```
mvn clean verify jacoco:report-aggregate
```
you _do_ get coverage, alas, the coverage of the **previous build**, because `clean` is _also_ integrated into the lifecycle. So when `report-aggregate` is executed on the parent, it will find the old `*.exec` files of the child modules, because they have not been cleaned, yet.

A dedicated aggregator-mojo OTOH will be executed after the lifecycle:
```
mvn clean verify jacoco:report-aggregate-all
```
This would also remove the need for the `onlyReactorProjects` property.

olamy commented:
@cpfeiffer Sorry, I forgot this one. The changes should be better.

cpfeiffer commented:
Awesome! Would it be good practice to let `ReportAggregateAllMojo` inherit `ReportAggregateMojo` and just overriding the `findDependencies()` (to be made protected then)?

olamy commented:
> Awesome! Would it be good practice to let `ReportAggregateAllMojo` inherit `ReportAggregateMojo` and just overriding the `findDependencies()` (to be made protected then)?

Oh right. good catch, 2 ezy. Just pushed it.

cpfeiffer commented:
Excellent, thanks a lot! :heart: 
So what do the the jacoco maintainers think?

olamy commented:
@Godin ping :)

olamy commented:
sadly, this doesn't look to attract any attention from the maintainer of the jacoco plugin.
Not sure why as it's real missing feature.
I will have to create a separate Maven plugin with only this single mojo. BTW a bit different as I will not be able to use exact same code which extend a mojo from here.

cpfeiffer commented:
> sadly, this doesn't look to attract any attention from the maintainer of the jacoco plugin. Not sure why, as it's a real missing feature. I will have to create a separate Maven plugin with only this single mojo. BTW a bit different as I will not be able to use exact same code which extends a mojo from here.

I hope it's just a time-issue.
This fix would make a [whole wiki-section](https://github.com/jacoco/jacoco/wiki/MavenMultiModule) of discussion/workarounds obsolete. I would be willing to update it. What do you think,  @Godin ?

olamy commented:
Maybe a timing issue for an idea/PR started 6 months ago...

olamy commented:
@cpfeiffer If you are interested, I have done this https://github.com/olamy/jacoco-aggregator-maven-plugin/ 
It's not the best solution, but after more than six months and no comment from here, I guess there is no interest. 
I will leave the PR open just in case.

marchof commented:
Dear @olamy, thanks for your contribution and very sorry for the silence from us maintainers. We (especially @Godin) mostly spend our spare time in keeping the core up-to-date. Due to our limited resources we're very reluctant to add new features which requires even more maintenance. Especially maintaining the Maven plugin is very exhausting as many community members have a strong opinion how things are done (at least in their context). We were already at the point to give-up on the integrations so that that the community can start separate projects.

Questions: Can you imagine to help us with the maintenance (bug fixing, feature/bug triage etc) for the Maven goals in the long run?

olamy commented:
@marchof, thanks for your comment and for answering a big part of my frustration :) 
No worries. I perfectly understand you may have limited time for maintenance.
I am happy to help maintain the Maven plugin part, as I may have some experience doing so.
Thanks

olamy commented:
@marchof @Godin thanks guys for your help and merge it for the last release 0.8.13.
very good community support here.

kwin commented:
FTR: This is available from https://github.com/olamy/jacoco-aggregator-maven-plugin. @olamy Thanks a lot!

cpfeiffer commented:
I have rebased this PR on master and adjusted the doc as requested here https://github.com/jacoco/jacoco/pull/1692#discussion_r2221470846

Since I can't push to this PR, shall I open a new one?

olamy commented:
> I have rebased this PR on master and adjusted the doc as requested here [#1692 (comment)](https://github.com/jacoco/jacoco/pull/1692#discussion_r2221470846)
> 
> Since I can't push to this PR, shall I open a new one?

I have added you as collaborator of the fork.
